pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/PinotHelixResourceManager.java

```diff
@@ -1559,33 +1560,32 @@ public void addTable(TableConfig tableConfig)
     if (ZKMetadataProvider.getSchema(_propertyStore, TableNameBuilder.extractRawTableName(tableNameWithType)) == null) {
       throw new InvalidTableConfigException("No schema defined for table: " + tableNameWithType);
     }
-    if (tableType == TableType.OFFLINE) {
-      try {
-        // Add table config
-        ZKMetadataProvider.setTableConfig(_propertyStore, tableNameWithType, TableConfigUtils.toZNRecord(tableConfig));
-        // Assign instances
-        assignInstances(tableConfig, true);
+    Preconditions.checkState(tableType == TableType.OFFLINE || tableType == TableType.REALTIME,
+        "Invalid table type: %s", tableType);
+    try {
+      // Add table config
+      if (!ZKMetadataProvider.createTableConfig(_propertyStore, tableNameWithType, tableConfig)) {
+        if (getTableConfig(tableNameWithType) != null) {
+          throw new TableAlreadyExistsException("Table config for " + tableNameWithType
+              + " already exists. If this is unexpected, try deleting the table to remove all metadata associated"
+              + " with it.");
+        } else {
+          throw new RuntimeException("Failed to create table config for table: " + tableNameWithType);
+        }
+      }
+      // Assign instances
+      assignInstances(tableConfig, true);
+      if (tableType == TableType.OFFLINE) {
         // Add ideal state
         _helixAdmin.addResource(_helixClusterName, tableNameWithType, idealState);
-      } catch (Exception e) {
-        LOGGER.error("Caught exception during offline table setup. Cleaning up table {}", tableNameWithType, e);
-        deleteOfflineTable(tableNameWithType);
-        throw e;
-      }
-    } else {
-      Preconditions.checkState(tableType == TableType.REALTIME, "Invalid table type: %s", tableType);
-      try {
-        // Add table config
-        ZKMetadataProvider.setTableConfig(_propertyStore, tableNameWithType, TableConfigUtils.toZNRecord(tableConfig));
-        // Assign instances
-        assignInstances(tableConfig, true);
+      } else {
         // Add ideal state with the first CONSUMING segment
         _pinotLLCRealtimeSegmentManager.setUpNewTable(tableConfig, idealState);
-      } catch (Exception e) {
-        LOGGER.error("Caught exception during realtime table setup. Cleaning up table {}", tableNameWithType, e);
-        deleteRealtimeTable(tableNameWithType);
-        throw e;
       }
+    } catch (Exception e) {
+      LOGGER.error("Caught exception during offline table setup. Cleaning up table {}", tableNameWithType, e);
+      deleteTable(tableNameWithType, tableType, null);
```

Comment:
This is risky. If 2 calls to this API happens at the same time, the just created table can be deleted here. We should probably move the `try` after the table config is posted

Reply:
Moved it.

I think technically we should have a cluster level lock to ensure certain options won't take place simultaneously.